README.rst
Introduction
============

``ftw.workspace`` provides a full featured project folder for Plone.

It uses ``ftw.tabbedview`` and ``ftw.activity`` to give you a overview view what happens
in your project.

What could happen in your project you're asking?

- Share files
    - Multi-upload using `collective.quickupload` - You are able to directly drop files into the Workspace, the Quickupload portlet is not necessary.
    - `ftw.file` provides a journal, versions, D'n'D file replacement, and much more.
    - Seperate Tabs, which shows all files, including filtering and Image preview.
- Manage your Team.
    - Extend your Team with ``ftw.participation``. Invite external an internal people to collaborate.
    - Overview of who is collaborating and open invitations.
- Meetings and Events
    - Add Meetings or Events.
    - Overview of all Meetings and Events as list or FullCalendar using ``ftw.calendar``.
- General features
    - Download all necessary informations of the Workspace as PDF.
    - Once you're done with the Project you can Download everything as ZIP-File.
    - Shows activities (Add, Modify, Delete, etc.)
    - Uploaded something interesting? Notify the People you want with ``ftw.notification``.
    - All Informations are fast accessible thru Tabs.
    - Create Project structures with Folders.
    - Global Overview of all your Projects with the ``@@workspaces_view``.
    - Using ExtJS for tabular listings.
    - Move and delete files directly in listings.



Integrated AddOns
-----------------
**``egov.contactdirectory``** is implemented as `extras`::


    [instance]
    eggs =+
        ftw.workspace [contact]


An additional profile is implemented which adds a new Tab to the Workspace and make the Contacts addable.


**PDF generation** is implemented as `extras`::

    [instance]
    eggs =+
        ftw.workspace [pdf]

Consider not installing it, if PDFLatex binaries are missing.


**ZIP Extraction**:
The package integrates well with `ftw.zipextract`_, making it possbile to extract
an uploaded ZIP file on the server::

    [instance]
    eggs =+
        ftw.workspace [zip_extract]


Possible AddOns
---------------
You may extend the Workspace by the following features.

- ``izug.ticketbox``
- ``ftw.mail``
- ``ftw.book``
- ``ftw.blog``
- ``ftw.contentpage``
- ``ftw.labels``
- ``ftw.downloadtoken``
- ``ftw.quota``
- ``ftw.avatar``



Example Workflow
================
``ftw.workspace`` is shipped with a ``ftw.lawgiver`` based default workflow for Workspaces.
You need to copy the workflow specifications of all three workflows (incl. placeful workflow policy) to your policy package and generate the necessary definition.xml using ftw.lawgiver. Check ``ftw.lawgiver`` for further informations.
You may also update the placeful workflow mappings with other types used in Workspaces.


Compatibility
=============

``ftw.workspace`` Version 2.0.0 and greater only supports Plone 4.3.x

Use ``ftw.workspace`` Version 1.7.x for Plone 4.1 and Plone 4.2.


Usage
=====

- Install the package by adding it to your buildout configuration:

::

    [instance]
    eggs =+
        ftw.workspace


- Install the generic setup profile.


Links
=====

- Github: https://github.com/4teamwork/ftw.workspace
- Issues: https://github.com/4teamwork/ftw.workspace/issues
- Pypi: http://pypi.python.org/pypi/ftw.workspace
- Continuous integration: https://jenkins.4teamwork.ch/search?q=ftw.workspace


Copyright
---------

This package is copyright by `4teamwork <http://www.4teamwork.ch/>`_.

``ftw.workspace`` is licensed under GNU General Public License, version 2.


.. _ftw.meeting: https://github.com/4teamwork/ftw.meeting
.. _ftw.zipextract: https://github.com/4teamwork/ftw.zipextract


docs/HISTORY.txt
Changelog
=========


4.3.1 (unreleased)
------------------

- Update workflows due to a new permission in `ftw.file 1.16.0`. [mbaechtold]


4.3.0 (2018-06-21)
------------------

- Add support for ftw.zipexport: configure types to create. [njohner]
- Exclude owner from assignable users vocabulary. [mbaechtold]


4.2.0 (2016-11-18)
------------------

- Speed up file listing rendering.
  The html2latex conversion was too slow for workspaces with a huge amount of files.
  This commit skips the html step and generates latex directly from template.
  [lknoepfel]

- Make setting roles on workspace creation more robust. [jone]


4.1.1 (2016-08-16)
------------------

- Add basetest to compare the workflow with the policyworkflow.
  [elioschmutz]


4.1.0 (2016-08-12)
------------------

- Update workspace workflows for the new requirements set by the workspace template in bobtemplates.
  [raphael-s]


4.0.1 (2016-08-11)
------------------

- Fix builders import.
  Do no longer import the egov.contactdirectory builders from the builders.py.
  If you want to use the ftw.workspace builders from another project, you don't
  want to import the egov.contactdirectory builders.
  [elioschmutz]


4.0.0 (2016-08-08)
------------------

- Implement example workspaces workflow.
  [mathias.leimgruber]

- Move several features from the main policy package to ``ftw.workspace``. Check
  Readme for details. If you upgrading to this version be careful, it may enable features you don't want.
  [mathias.leimgruber]


3.1.1 (2015-06-03)
------------------

- Fix ftw.pdfgenerator import without dependency.
  [jone]


3.1.0 (2015-05-26)
------------------

- Add a spreadsheet of the participants in the zip export.
  [mbaechtold]

- Export workspace as a ZIP file with the help of ftw.zipexport.
  [mbaechtold]

- Add the fullname of the creator of a workspace to the searchable text
  index. This way the list of workspaces on the overview tab can be filtered
  by creator.
  [mbaechtold]


3.0.1 (2014-11-19)
------------------

- Do not show filter/search box on activitiy tab.


3.0.0 (2014-09-29)
------------------

- Add "Activity"-tab to the workspaces view.
  [jone]

- Overview: replace old recently modified with activity stream.
  [jone]


2.2.2 (2015-04-13)
------------------

- Add the fullname of the creator of a workspace to the searchable text
  index. This way the list of workspaces on the overview tab can be filtered
  by creator.
  [mbaechtold]


2.2.1 (2014-09-03)
------------------

- Use navtree properties to sort items on workspace overview.
  [tschanzt]

- Fix broken upgradestep 1707 if no workspaces exists.
  [elioschmutz]


2.2.0 (2014-06-02)
------------------

- Added new helper method to get icon or sprite for document tab.
  [Julian Infanger]

- Show all users with local roles on a Workspace.
  [mathias.leimgruber]

- Hide upload option of text field on Workspace and TabbedViewFolder.
  [mathias.leimgruber]

- Fix misplaced checkbox in documents tab. Reset table configuration.
  [elioschmutz]

- Use ftw.builder ImageBuilder, remove the ImageBuilder of this package.
  [mathias.leimgruber]


2.1.0 (2014-03-20)
------------------

- Quick upload factory: also reindex dexterity types.
  [jone]

- Add functionality for deleting and moving multiple documents
  directly from the documents tab of the workspace.
  [lknoepfel, elioschmutz]


2.0.4 (2014-02-05)
------------------

- Remove old, no longer used calendar js code.
  [buchi]

- Update German, French and English translations.
  [jone]


2.0.3 (2014-01-09)
------------------

- Fix quick_upload id generator (allow filenames staring with an '_')
  [mathias.leimgruber]


2.0.2 (2013-11-28)
------------------

- Add txt preview adapter (shows just a txt.png).
  [mathias.leimgruber]

- Add zip preview adapter (shows just a zip.png).
  [mathias.leimgruber]

- Add pdf preview adapter (shows just a pdf.png).
  [mathias.leimgruber]

- Add xlsx preview adapter (shows just a xlsx.png).
  [mathias.leimgruber]

- Add xls preview adapter (shows just a xls.png).
  [mathias.leimgruber]

- Add pptx preview adapter (shows just a pptx.png).
  [mathias.leimgruber]

- Add ppt preview adapter (shows just a ppt.png).
  [mathias.leimgruber]

- Add docx preview adapter (shows just a docx.png).
  [mathias.leimgruber]

- Add doc preview adapter (shows just a doc.png).
  [mathias.leimgruber]

- Add new fallback image for previews.
  [mathias.leimgruber]


2.0.1 (2013-11-15)
------------------

- The preview image should not be bigger than the container.
  [Julian Infanger]


2.0.0 (2013-11-04)
------------------

- Group previews by modification date.
  [mathias.leimgruber]

- Implement image preview for ftw.file.
  [mathias.leimgruber]


1.8.1 (2013-11-01)
------------------

- Fix bad href in overview_sublisting.
  [elioschmutz]


1.8.0 (2013-10-24)
------------------

- Implement a generic listing on the workspace overview tab.
  [mathias.leimgruber]


1.7.1 (2013-09-16)
------------------

- Use human date formatting with "today", "yesterday" etc.
  for tabs "documents", "events" and "workspaces".
  [jone]


1.7.0 (2013-08-28)
------------------

- Add div-element arount ics export for better css selectors
  [elio.schmutz]

- Provide a workspace browserlayer.
  [mathias.leimgruber]

- Make auto-roles, granted to the owner of a workspace, configurable.
  [jone]

- Plone 4.3 compatibility.
  [jone]

- Move the following features from a private repo to ftw.workspace.

  - Enable filtering on overview.
  - Truncate long text in recently modified listing.
  - Use 'today + time' and 'yesterday + time' in recently modified listing.

  [mathias.leimgruber]

- Add permissionmapping for ftw.lawgiver
  [elioschmutz]


1.6.3 (2013-03-11)
------------------

- Make pdf export work with document_date.
  [tschanzt]

- Use finalize schemata also for TabbedViewFolder.
  [mathias.leimgruber]

- Replace jq by $.
  [mathias.leimgruber]

1.6.2 (2013-01-21)
------------------


ftw/workspace/patches.py
from Products.CMFCalendar.CalendarTool import CalendarTool
from Products.CMFCore.utils import getToolByName

def ftw_getEventsForCalendar(self, month='1', year='2002', path=''):
    """ recreates a sequence of weeks, by days each day is a mapping.
        {'day': #, 'url': None}
    """
    year = int(year)
    month = int(month)
    # daysByWeek is a list of days inside a list of weeks, like so:
    # [[0, 1, 2, 3, 4, 5, 6],
    #  [7, 8, 9, 10, 11, 12, 13],
    #  [14, 15, 16, 17, 18, 19, 20],
    #  [21, 22, 23, 24, 25, 26, 27],
    #  [28, 29, 30, 31, 0, 0, 0]]
    daysByWeek = self._getCalendar().monthcalendar(year, month)
    weeks = []

    events = self.catalog_getevents(year, month,  path=path)

    for week in daysByWeek:
        days = []
        for day in week:
            if events.has_key(day):
                days.append(events[day])
            else:
                days.append({'day': day, 'event': 0, 'eventslist':[]})

        weeks.append(days)

    return weeks

def ftw_catalog_getevents(self, year, month, path=''):
    """ given a year and month return a list of days that have events 
    """
    # XXX: this method violates the rules for tools/utilities:
    # it depends on a non-utility tool
    year = int(year)
    month = int(month)
    last_day = self._getCalendar().monthrange(year, month)[1]
    first_date = self.getBeginAndEndTimes(1, month, year)[0]
    last_date = self.getBeginAndEndTimes(last_day, month, year)[1]

    ctool = getToolByName(self, 'portal_catalog')
    query = ctool(
                    portal_type=self.getCalendarTypes(),
                    review_state=self.getCalendarStates(),
                    start={'query': last_date, 'range': 'max'},
                    end={'query': first_date, 'range': 'min'},
					path=path,
                    sort_on='start' )

    # compile a list of the days that have events
    eventDays={}
    for daynumber in range(1, 32): # 1 to 31
        eventDays[daynumber] = {'eventslist': [],
                                'event': 0,
                                'day': daynumber}
    includedevents = []
    for result in query:
        if result.getRID() in includedevents:
            break
        else:
            includedevents.append(result.getRID())
        event={}
        # we need to deal with events that end next month
        if  result.end.month() != month:
            # doesn't work for events that last ~12 months
            # fix it if it's a problem, otherwise ignore
            eventEndDay = last_day
            event['end'] = None
        else:
            eventEndDay = result.end.day()
            event['end'] = result.end.Time()
        # and events that started last month
        if result.start.month() != month:  # same as above (12 month thing)
            eventStartDay = 1
            event['start'] = None
        else:
            eventStartDay = result.start.day()
            event['start'] = result.start.Time()

        event['title'] = result.Title or result.getId

        if eventStartDay != eventEndDay:
            allEventDays = range(eventStartDay, eventEndDay+1)
            eventDays[eventStartDay]['eventslist'].append(
                    {'end': None,
                     'start': result.start.Time(),
                     'title': event['title']} )
            eventDays[eventStartDay]['event'] = 1

            for eventday in allEventDays[1:-1]:
                eventDays[eventday]['eventslist'].append(
                    {'end': None,
                     'start': None,
                     'title': event['title']} )
                eventDays[eventday]['event'] = 1

            if result.end == result.end.earliestTime():
                last_day_data = eventDays[allEventDays[-2]]
                last_days_event = last_day_data['eventslist'][-1]
                last_days_event['end'] = (result.end-1).latestTime().Time()
            else:
                eventDays[eventEndDay]['eventslist'].append( 
                    { 'end': result.end.Time()
                    , 'start': None, 'title': event['title']} )
                eventDays[eventEndDay]['event'] = 1
        else:
            eventDays[eventStartDay]['eventslist'].append(event)
            eventDays[eventStartDay]['event'] = 1
        # This list is not uniqued and isn't sorted
        # uniquing and sorting only wastes time
        # and in this example we don't need to because
        # later we are going to do an 'if 2 in eventDays'
        # so the order is not important.
        # example:  [23, 28, 29, 30, 31, 23]
    return eventDays

CalendarTool.catalog_getevents = ftw_catalog_getevents
CalendarTool.getEventsForCalendar = ftw_getEventsForCalendar